client/src/document/toolbar/flaws.tsx

```diff
@@ -107,61 +109,93 @@ function Flaws({ flaws }: { flaws: FlawCheck[] }) {
+  //
+  // So when you loop over `links` you'll know many times the `link.href`
+  // has appeared before.
+  // Now, when you've filtered the list of
+  // all `document.querySelectorAll("a[href]")` that matches 'foo'
+  // you know *which* one to pick.
+  const indexes = {};
+  const nths: number[] = [];
+  for (const link of links) {
+    if (!(link.href in indexes)) {
+      indexes[link.href] = 0;
     }
-  );
+    nths.push(indexes[link.href]);
+    indexes[link.href]++;
+  }
 
+  const [opening, setOpening] = React.useState<string | null>(null);
   useEffect(() => {
-    // 'data' will be 'undefined' until the server has had a chance to
-    // compute all the redirects. Don't bother until we have that data.
-    if (!data) {
-      return;
+    let unsetOpeningTimer: ReturnType<typeof setTimeout>;
+    if (opening) {
+      unsetOpeningTimer = setTimeout(() => {
+        setOpening(null);
+      }, 3000);
     }
+    return () => {
+      if (unsetOpeningTimer) {
+        clearTimeout(unsetOpeningTimer);
+      }
+    };
+  }, [opening]);
+
+  const filepath = sourceFolder + "/index.html";
+
+  function openInEditor(key: string, line: number, column: number) {
+    const sp = new URLSearchParams();
+    sp.set("filepath", filepath);
+    sp.set("line", `${line}`);
+    sp.set("column", `${column}`);
+    console.log(
+      `Going to try to open ${filepath}:${line}:${column} in your editor`
+    );
+    setOpening(key);
+    fetch(`/_open?${sp.toString()}`).catch((err) => {
+      console.warn(`Error trying to _open?${sp.toString()}:`, err);
+    });
+  }
+
+  useEffect(() => {
     const annotations: RoughAnnotation[] = [];
     // If the anchor already had a title, put it into this map.
     // That way, when we restore the titles, we know what it used to be.
-    const originalTitles = new Map();
-    for (const anchor of [
-      ...document.querySelectorAll<HTMLAnchorElement>("div.content a[href]"),
-    ]) {
-      const hrefURL = new URL(anchor.href);
-      const { pathname, hash } = hrefURL;
-      if (pathname in data.redirects) {
-        // Trouble! But is it a redirect?
-        let correctURI = data.redirects[pathname];
-        let annotationColor = "red";
-        originalTitles.set(anchor.href, anchor.title);
-        if (correctURI) {
-          if (hash) {
-            correctURI += hash;
-          }
-          // It can be fixed!
-          annotationColor = "orange";
-          anchor.title = `Consider fixing! It's actually a redirect to ${correctURI}`;
-        } else {
-          anchor.title = "Broken link! Links to a page that will not be found";
-        }
+    links.forEach((link, i) => {
+      const matchedAnchors = [
+        ...document.querySelectorAll<HTMLAnchorElement>("div.content a[href]"),
+      ].filter(
+        (anchor) =>
+          anchor.href === link.href ||
+          new URL(anchor.href).pathname === link.href
+      );
+      const anchor = matchedAnchors[nths[i]];
+      if (anchor) {
```

Comment:
this things feels a bit backwards to me, I think it would be more readable if we'd flip it, i.e. something like:

```jsx
let linkIndex = 0;
for (const anchor in [...document.querySelectorAll<HTMLAnchorElement>("div.content a[href]")]) {
  const link = links[linkIndex];
  if (!link) {
    break;
  }
  if (anchor.href !== link.href && new URL(anchor.href).pathname !== link.href) {
    continue;
  }

  doTheAnchorAnnotating();
}
```

I feel like I might be missing something here, because it seems to alleviate the need for `indexes` and `nths` but that might be due to me overlooking something?

Reply:
This is why code reviews are so good and useful. Yeah, your algorithm *is* better and it works. I've now copied it.